ring_theory/localization.lean

```diff
@@ -0,0 +1,263 @@
+{ intros,
+  try {cases a with r₁ s₁, cases s₁ with s₁ hs₁},
+  try {cases b with r₂ s₂, cases s₂ with s₂ hs₂},
+  try {cases c with r₃ s₃, cases s₃ with s₃ hs₃},
+  apply quotient.sound,
+  existsi (1:α),
+  existsi is_submonoid.one_mem S,
+  simp [mul_left_comm, mul_add, mul_comm] }
+
+def of_comm_ring : α → loc α S :=
+λ r, ⟦⟨r, 1, is_submonoid.one_mem S⟩⟧
+
+instance : is_ring_hom (of_comm_ring α S) :=
+{ map_add := λ x y, quotient.sound $ by simp,
+  map_mul := λ x y, quotient.sound $ by simp,
+  map_one := rfl }
+
+local infix ^ := monoid.pow
+
+variable {α}
+
+def away (x : α) := loc α (powers x)
+
+instance away.comm_ring (x : α) : comm_ring (away x) :=
+localization.comm_ring α (powers x)
+
+section at_prime
+
+variables (P : set α) [is_prime_ideal P]
+
+instance prime.is_submonoid :
+  is_submonoid α (set.compl P) :=
+{ one_mem := λ h, is_prime_ideal.ne_univ P $
+    is_submodule.univ_of_one_mem P h,
+  mul_mem := λ x y hnx hny hxy, or.cases_on
+    (is_prime_ideal.mem_or_mem_of_mul_mem hxy) hnx hny }
+
+def at_prime := loc α (set.compl P)
+
+instance at_prime.comm_ring : comm_ring (at_prime P) :=
+localization.comm_ring α (set.compl P)
+
+instance at_prime.local_ring : local_ring (at_prime P) :=
+local_of_nonunits_ideal
+  (λ hze, have _, from quotient.exact hze, let ⟨t, hts, ht⟩ := this in
+     hts $ have htz : t = 0, by simpa using ht,
+       suffices (0:α) ∈ P, by rwa htz,
+       @is_submodule.zero _ _ _ _ P _)
+  (λ x y hx hy ⟨z, hz⟩,
+     let ⟨⟨r₁, s₁, hs₁⟩, hrs₁⟩ := quotient.exists_rep x,
+         ⟨⟨r₂, s₂, hs₂⟩, hrs₂⟩ := quotient.exists_rep y,
+         ⟨⟨r₃, s₃, hs₃⟩, hrs₃⟩ := quotient.exists_rep z in
+     have _, by rw [← hrs₁, ← hrs₂, ← hrs₃] at hz; from quotient.exact hz,
+     let ⟨t, hts, ht⟩ := this in
+     have hr₁ : r₁ ∈ P, from classical.by_contradiction $
+     λ hnr₁, hx ⟨⟦⟨s₁, r₁, hnr₁⟩⟧, by rw ←hrs₁;
+       from quotient.sound ⟨1, is_submonoid.one_mem _, by simp [mul_comm]⟩⟩,
+     have hr₂ : r₂ ∈ P, from classical.by_contradiction $
+     λ hnr₂, hy ⟨⟦⟨s₂, r₂, hnr₂⟩⟧, by rw ←hrs₂;
+       from quotient.sound ⟨1, is_submonoid.one_mem _, by simp [mul_comm]⟩⟩,
+     have hr₃ : _ , from or.resolve_right
+     (mem_or_mem_of_mul_eq_zero P ht) hts,
+     have h : s₃ * (s₁ * s₂) - r₃ * (s₁ * r₂ + s₂ * r₁) ∈ P,
+     by simpa using hr₃,
+     have h1 : r₃ * (s₁ * r₂ + s₂ * r₁) ∈ P,
+     from is_submodule.smul r₃ $ is_submodule.add
+         (is_submodule.smul s₁ hr₂)
+         (is_submodule.smul s₂ hr₁),
+     have h2 : s₃ * (s₁ * s₂) ∈ P,
+     from calc s₃ * (s₁ * s₂)
+           = s₃ * (s₁ * s₂) - r₃ * (s₁ * r₂ + s₂ * r₁) + r₃ * (s₁ * r₂ + s₂ * r₁) : eq.symm $ sub_add_cancel _ _
+       ... ∈ P : is_submodule.add h h1,
+     have h3 : s₁ * s₂ ∈ P, from or.resolve_left
+     (is_prime_ideal.mem_or_mem_of_mul_mem h2) hs₃,
+     or.cases_on (is_prime_ideal.mem_or_mem_of_mul_mem h3) hs₁ hs₂)
+
+end at_prime
+
+def closure (S : set α) : set α := {y | ∃ (L:list α) (H:∀ x ∈ L, x ∈ S), L.prod = y }
```

Comment:
`∃(L:list α), (∀ x ∈ L, x ∈ S) ∧ L.prod = y `

You know that you also have the option to use an inductive for closure:

```lean
inductive in_closure {α : Type} [monoid α]  (S : set α) : α → Prop
| basic : ∀a∈S, in_closure a
| one : in_closure 1
| mul : ∀x y, in_closure x → in_closure y → in_closure (x * y)

def closure (S : set α) : set α := {x | in_closure x}
instance closure.is_submonoid (S : set α) : is_submonoid α (closure S) :=
{ one_mem := in_clousure.one, mul_mem := in_closure.mul } 
```

Reply:
That's really interesting to see how you and Mario think. I had the same discussion with Mario when I was playing with generating sets for groups. He convinced me to try the inductive definition. Every proof became complicated and, more importantly, unrecognizable. There was nothing I would have written on a blackboard.

I'm sure there is good reason to suggest this, I only wish I could understand.

Reply:
@PatrickMassot I would like to refute your claim. Do you have any particular examples where we can try things out in multiple ways?

I know that mathematicians aren't used to having inductively defined predicates as "primitive", and so they tend to prefer descriptions that involve sequences or "..." in them. But the ability to have arbitrarily complicated recursion principles on demand is one of the more powerful features of lean's axiomatics, and it generally saves a lot of work.

Reply:
I think Patrick is simply pointing out that this whole way of thinking about mathematics on the whole takes quite some getting used to. I look at the two definitions proposed and they are manifestly "the same" for some mathematical interpretation of this phrase. Perhaps all that is going on is that the inductive definition is simply second nature to a mathematician and we don't really make a fuss about it because mathematics students are not trained to be picky about these matters. Perhaps there are two kinds of proofs that one would see in practice. There are those where a mathematician would write "let y=x1*x2*...*xn be a general element of the closure" and then somehow use y directly and not ever really care about n. Here is where the non-inductive definition looks very natural (to me) and the inductive one looks rather clunky in comparison. And then there are those where we would write "let y=x1*x2*...*xn be a general element of the closure; by a trivial induction on n we see [blah]". Here I would imagine that the inductive definition might well work better -- but we just want to argue by induction on length of list (which I could certainly imagine happens in practice) because to a mathematician the main inductive principle which is used 99 percent of the time is induction on the naturals, so we locate the natural (here list length) and induct on that rather than inducting on anything more exotic (what is the point?). However I am wondering whether mathematicians are in fact misguided here, and what the CS guys are pointing out is that actually in practice 99 percent of things, when proved formally, are most naturally proved by induction over the more complex structure, and most of the time we just don't notice because induction on n looks like it works fine on a more informal level. It would be interesting to see test cases where induction on list length actually turns out to be a pain and induction on the inductive type works more painlessly, but the proof of the pudding is in the eating as we say in the UK...

Reply:
In the end we want to have both definitions, maybe using multisets instead of lists (when the monoid is commutative). But the advantage of starting with the inductive definition is that we can use the equation compiler. 

There is also an alternative inductive definition:
```lean
inductive in_closure (S : set α) : α → Prop
| one : in_closure 1
| mul_basic : ∀x∈S, ∀y, in_closure x → in_closure (x * y)
```

But finally we want to have the inductive view, want to know that each element of the closure is generated by a multiset / list, and we want to see the closure as the smallest submonoid containing `S`. From a abstract view it doesn't matter where we start and end. But due to the equation compiler we may want to start with a inductive definition.

Reply:
The specific example where we discussed that together is my early attempts at defining norms on non-commutative groups from generating sets. You can see the result with lists on https://github.com/PatrickMassot/lean-scratchpad/blob/master/norms.lean

It looks like what I started with back in https://gitter.im/leanprover_public/Lobby?at=5a35833a0163b0281045ce0f
but if you read gitter from this point you see Mario telling me to switch to something like:
https://gitter.im/leanprover_public/Lobby?at=5a359a8b5355812e57e3ee2b
and even getting confused about the correct definition. Then I tried to prove stuff using this and everything was horrible. And I switched back to lists.